runtime/compiler/net/ClientStream.hpp

```diff
@@ -0,0 +1,223 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef CLIENT_STREAM_H
+#define CLIENT_STREAM_H
+
+#include <google/protobuf/io/zero_copy_stream_impl.h>
+#include "compile/CompilationTypes.hpp"
+#include "ilgen/J9IlGeneratorMethodDetails.hpp"
+#include "net/ProtobufTypeConvert.hpp"
+#include "net/CommunicationStream.hpp"
+
+#if defined(JITSERVER_ENABLE_SSL)
+#include <openssl/ssl.h>
+class SSLOutputStream;
+class SSLInputStream;
+#endif
+
+namespace JITServer
+{
+enum VersionCheckStatus
+   {
+   NOT_DONE = 0,
+   PASSED = 1,
+   };
+
+/**
+   @class ClientStream
+   @brief Implementation of the communication API for a client asking for remote JIT compilations
+
+   Typical sequence executed by a client is:
+   (1) Establish a connection to the JITServer by creating a ClientStream object
+         client = new (PERSISTENT_NEW) JITServer::ClientStream(compInfo->getPersistentInfo());
+   (2) Determine compilation parameters and send the compilation request over the network
+          client->buildCompileRequest(method, clazz, .................);
+   (3) Handle any queries from the server
+       Typical seq: (3.1) client->read(); (3.2) client->getRecvData(...); (3.3) client->write(...);
+          while(!handleServerMessage(client, frontend));
+   (4) Read compilation result
+          auto recv = client->getRecvData<uint32_t, std::string, std::string, .......>();
+   (5) install compiled code into code cache
+*/
+class ClientStream : CommunicationStream
```

Comment:
I don't think it should be `private` inheritance here since it's not exactly a has-a relationship between `ClientStream`  and `CommunicationStream`.  Will change it to `public`.